File: SDK/ios/Classes/listener/AudioFrameListener.swift
//
//  AudioFrameListener.swift
//  tencent_trtc_cloud
//
//  Created by vincepzhang on 2023/12/26.
//

import Foundation
import TXLiteAVSDK_Professional

class AudioFrameListener: NSObject, TRTCAudioFrameDelegate {
    let basicChannel: FlutterBasicMessageChannel
    
    init(basicChannel: FlutterBasicMessageChannel) {
        self.basicChannel = basicChannel
    }
    
    func onCapturedAudioFrame(_ frame: TRTCAudioFrame) {
        var dataByteArray: [UInt8] = []
        frame.data.withUnsafeBytes { (pointer: UnsafeRawBufferPointer) in
            dataByteArray = Array(pointer)
        }
        
        var extraDataByteArray: [UInt8] = []
        frame.extraData?.withUnsafeBytes { (pointer: UnsafeRawBufferPointer) in
            extraDataByteArray = Array(pointer)
        }
        
        basicChannel.sendMessage(["method": "onCapturedAudioFrame",
                                  "params": ["data": dataByteArray,
                                             "sampleRate": frame.sampleRate.rawValue,
                                             "channels": frame.channels,
                                             "timestamp": frame.timestamp,
                                             "extraData": extraDataByteArray]])
    }
    
    func onLocalProcessedAudioFrame(_ frame: TRTCAudioFrame) {
        // MARK: TODO
    }
    
    func onRemoteUserAudioFrame(_ frame: TRTCAudioFrame, userId: String) {
        // MARK: TODO
    }
    
    func onMixedAllAudioFrame(_ frame: TRTCAudioFrame) {
        // MARK: TODO
    }
    
    func onMixedPlay(_ frame: TRTCAudioFrame) {
        // MARK: TODO
    }
    
    func onVoiceEarMonitorAudioFrame(_ frame: TRTCAudioFrame) {
        // MARK: TODO
    }
}

File: SDK/android/src/main/java/com/tencent/trtcplugin/listener/AudioFrameListener.java
package com.tencent.trtcplugin.listener;

import android.os.Handler;
import android.os.Looper;

import com.tencent.trtc.TRTCCloudDef;
import com.tencent.trtc.TRTCCloudListener;

import java.util.HashMap;
import java.util.Map;

import io.flutter.plugin.common.BasicMessageChannel;

public class AudioFrameListener implements TRTCCloudListener.TRTCAudioFrameListener {
    private BasicMessageChannel mBasicChannel;
    private Handler             mUIHandler;

    public AudioFrameListener(BasicMessageChannel basicChannel) {
        this.mBasicChannel = basicChannel;
        mUIHandler = new Handler(Looper.getMainLooper());
    }

    @Override
    public void onCapturedAudioFrame(TRTCCloudDef.TRTCAudioFrame trtcAudioFrame) {
        mUIHandler.post(() -> {
            Map<String, Object> params = new HashMap();
            params.put("data", trtcAudioFrame.data);
            params.put("sampleRate", trtcAudioFrame.sampleRate);
            params.put("channels", trtcAudioFrame.channel);
            params.put("timestamp", trtcAudioFrame.timestamp);
            params.put("extraData", trtcAudioFrame.extraData);

            Map<String, Object> message = new HashMap();
            message.put("method", "onCapturedAudioFrame");
            message.put("params", params);

            mBasicChannel.send(message);
        });
    }

    @Override
    public void onLocalProcessedAudioFrame(TRTCCloudDef.TRTCAudioFrame trtcAudioFrame) {
        //TODO
    }

    @Override
    public void onRemoteUserAudioFrame(TRTCCloudDef.TRTCAudioFrame trtcAudioFrame, String s) {
        //TODO
    }

    @Override
    public void onMixedPlayAudioFrame(TRTCCloudDef.TRTCAudioFrame trtcAudioFrame) {
        //TODO
    }

    @Override
    public void onMixedAllAudioFrame(TRTCCloudDef.TRTCAudioFrame trtcAudioFrame) {
        //TODO
    }

    @Override
    public void onVoiceEarMonitorAudioFrame(TRTCCloudDef.TRTCAudioFrame trtcAudioFrame) {
        //TODO
    }
}
